### PR TITLE
Add ability to exclude more resources via the cloud-nuke-excuded tag

### PR DESCRIPTION
**Pull request by merkata**
Currently, you can only exclude S3 buckets that are tagged with key cloud-nuke-excluded and value set to true. This PR adds more resources that are protected via this tag. This functionality is used in case you have a test environment that you do not want to get wiped out, like an integration setup with an EKS cluster, ELB, workers in an ASG etc. - you can propagate the tag across your setup for more fine-grained control (in comparison to specifying a creation time). The current setup is backwards compatible and has been tested with every resource.

### CODE REVIEW

**Comment by harmrai-pol** [comment]
Nice work @merkata! What is preventing this from getting merged in @brikis98 @yorinasub17

**Comment by gmembre-zenika** [comment]
It would  be great if this feature can be reviewed / merged 💚

**Comment by tommy31** [comment]
Bump.

This PR is related to this issue [166](https://github.com/gruntwork-io/cloud-nuke/issues/166)

**Comment by gforien** [comment]
Hi 👋 
Any updates on this ?

**Comment by ellisonc** [comment]
@merkata this change set looks great to me, could you resolve the conflicts in the PR and add test cases for the new tagging feature?

**Comment by merkata** [comment]
> @merkata this change set looks great to me, could you resolve the conflicts in the PR and add test cases for the new tagging feature?

Thanks a lot @ellisonc  - I've fixed the conflicts and I'll be writing the tests to cover the new functionality. Appreciate your input!

**Comment by zackproser** [comment]
> > @merkata this change set looks great to me, could you resolve the conflicts in the PR and add test cases for the new tagging feature?
> 
> Thanks a lot @ellisonc - I've fixed the conflicts and I'll be writing the tests to cover the new functionality. Appreciate your input!

@merkata This looks like a great change! Could you also please update the README to reflect this expanded functionality as well?

**Comment by merkata** [comment]
> > > @merkata this change set looks great to me, could you resolve the conflicts in the PR and add test cases for the new tagging feature?
> > 
> > 
> > Thanks a lot @ellisonc - I've fixed the conflicts and I'll be writing the tests to cover the new functionality. Appreciate your input!
> 
> @merkata This looks like a great change! Could you also please update the README to reflect this expanded functionality as well?

Thanks @zackproser , I'll update the README as well!

**Comment by ellisonc** [comment]
@merkata apologies for the delay, I attempted to fix the conflicts and get this merged but our tooling doesn't allow pushes to any branch named master and I don't want to steal your credit by making the changes on a new branch of my own.  Could you resolve the conflicts on a feature branch and update the PR?

**Comment by merkata** [comment]
> @merkata apologies for the delay, I attempted to fix the conflicts and get this merged but our tooling doesn't allow pushes to any branch named master and I don't want to steal your credit by making the changes on a new branch of my own. Could you resolve the conflicts on a feature branch and update the PR?

Hey @ellisonc , thanks for the feedback! I fixed the conflicts, two things stand out:

* the current `README.md` doesn't seem to cover exclusion of resources based on tags, the previous list of resources described which resources can be filtered out based on tag and I extended the descriptions of the resources I added.

* please have a look at `aws/snapshot.go`, I believe the logic is correct, still there is the addition of not deleting a snapshot with the AWS backup tag and I want to make sure I got the logic right.

Thanks!

File: aws/nat_gateway.go [hunk 1]
```diff
@@ -54,7 +54,11 @@ func shouldIncludeNatGateway(ngw *ec2.NatGateway, excludeAfter time.Time, config
 		return false
 	}
 
+<<<<<<< HEAD
+	if hasNGWExcludeTag(ngw) {
+=======
 	if aws.StringValue(ngw.State) == ec2.NatGatewayStateDeleted {
+>>>>>>> upstream/master
```
Comment (ellisonc) [inline aws/nat_gateway.go:61]: Weirdness from merge conflict here

**Review by ellisonc — COMMENTED** [review]
@merkata the logic all looks good, just some leftover issues from the merge conflict fix

**Comment by ellisonc** [comment]
@merkata I also don't see the readme update you mentioned, did that not get staged?

File: aws/nat_gateway.go (hunk 1 above)
Comment (merkata) [reply aws/nat_gateway.go]: Totally missed that one, sorry! Fixed.

**Comment by merkata** [comment]
> @merkata I also don't see the readme update you mentioned, did that not get staged?

I didn't explain it accurately - I updated the old version of the README.md, it looked like this:

```text
Inspecting and deleting all ACM Private CA in an AWS account
Inspecting and deleting all Auto scaling groups in an AWS account - except for ASGs tagged with Key=cloud-nuke-excluded Value=true
Inspecting and deleting all Elastic Load Balancers (v1 and v2) in an AWS account - except for ELBs tagged with Key=cloud-nuke-excluded Value=true
Inspecting and deleting all Transit Gateways in an AWS account
Inspecting and deleting all EBS Volumes in an AWS account - except for volumes tagged with Key=cloud-nuke-excluded Value=true
...
```

and I specifically added the _except for \<resource name\> tagged with..._ section for each resource.

Now the section looks like this:

```text
Cloud-nuke suppports 🔎 inspecting and 🔥💀 deleting the following AWS resources:

| Resource Family | Resource type 
| --------------- | ---------- 
| EC2 | Auto scaling groups |
| EC2 | Elastic Load Balancers (v1 and v2) |
| EC2 | EBS Volumes | 
| EC2 | Unprotected EC2 instances |
...
```

So I'm not sure where to make a note about excluding a resource based on an exclusion tag.

**Comment by merkata** [comment]
> @merkata I see what you're saying. How about adding a new section header under the [`Excluding Resources by Age`](https://github.com/gruntwork-io/cloud-nuke#excluding-resources-by-age) header (or maybe right above it?)

@zackproser sorry I wasn't able to response earlier, I've added a short paragraph with the list of resources that would be supported, let me know what you think of it. Thanks!

File: aws/asg.go [hunk 2]
```diff
@@ -32,6 +32,17 @@ func getAllAutoScalingGroups(session *session.Session, region string, excludeAft
 	return groupNames, nil
 }
 
+// hasASGExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
Comment (zackproser) [inline aws/asg.go:35]:
```suggestion
// hasASGExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

File: aws/ebs.go [hunk 3]
```diff
@@ -43,6 +43,17 @@ func getAllEbsVolumes(session *session.Session, region string, excludeAfter time
 	return volumeIds, nil
 }
 
+// hasEBSExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
Comment (zackproser) [inline aws/ebs.go:46]:
```suggestion
// hasEBSExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

File: aws/ebs.go [hunk 4]
```diff
@@ -43,6 +43,17 @@ func getAllEbsVolumes(session *session.Session, region string, excludeAfter time
...
+func hasEBSExcludeTag(volume *ec2.Volume) bool {
+	// Exclude deletion of any buckets with cloud-nuke-excluded tags
+	for _, tag := range volume.Tags {
+		if *tag.Key == AwsResourceExclusionTagKey && *tag.Value == "true" {
```
Comment (zackproser) [inline aws/ebs.go:50]: NIT - In general we want to use the `aws.StringValue` and related methods for dereferencing pointers - partly for consistency, but mainly for the fact that these methods include a safety check to ensure we're not dereferencing a nil pointer.

File: aws/eip.go [hunk 5]
```diff
@@ -82,6 +82,17 @@ func getAllEIPAddresses(session *session.Session, region string, excludeAfter ti
 	return allocationIds, nil
 }
 
+// hasEIPExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
Comment (zackproser) [inline aws/eip.go:85]:
```suggestion
// hasEIPExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

File: aws/elbv2.go [hunk 6]
```diff
@@ -18,22 +18,41 @@ import (
...
+		if shouldIncludeELBv2(balancer, excludeAfter, configObj, tags) {
 			arns = append(arns, balancer.LoadBalancerArn)
 		}
 	}
 
 	return arns, nil
 }
 
-func shouldIncludeELBv2(balancer *elbv2.LoadBalancer, excludeAfter time.Time, configObj config.Config) bool {
+// hasELBv2ExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
Comment (zackproser) [inline aws/elbv2.go:42]:
```suggestion
// hasELBv2ExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

File: aws/iam.go [hunk 7]
```diff
@@ -32,14 +32,27 @@ func getAllIamUsers(session *session.Session, excludeAfter time.Time, configObj
...
+			userNames = append(userNames, user.UserName)
+		} else if !hasIAMExcludeTag(user) {
 			userNames = append(userNames, user.UserName)
 		}
 	}
 
 	return userNames, nil
 }
 
+// hasIAMExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
Comment (zackproser) [inline aws/iam.go:45]:
```suggestion
// hasIAMExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

File: aws/nat_gateway.go [hunk 8]
```diff
@@ -65,6 +70,17 @@ func shouldIncludeNatGateway(ngw *ec2.NatGateway, excludeAfter time.Time, config
 	)
 }
 
+// hasNGWExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
Comment (zackproser) [inline aws/nat_gateway.go:73]:
```suggestion
// hasNGWExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

File: aws/rds.go [hunk 9]
```diff
@@ -34,6 +37,17 @@ func getAllRdsInstances(session *session.Session, excludeAfter time.Time, config
 	return names, nil
 }
 
+// hasRDSExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
Comment (zackproser) [inline aws/rds.go:40]:
```suggestion
// hasRDSExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

File: aws/rds_cluster.go [hunk 10]
```diff
@@ -45,14 +45,27 @@ func getAllRdsClusters(session *session.Session, excludeAfter time.Time) ([]*str
...
+			names = append(names, database.DBClusterIdentifier)
+		} else if !hasRDSClusterExcludeTag(database) {
 			names = append(names, database.DBClusterIdentifier)
 		}
 	}
 
 	return names, nil
 }
 
+// hasRDSClusterExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
Comment (zackproser) [inline aws/rds_cluster.go:58]:
```suggestion
// hasRDSClusterExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

File: aws/snapshot.go [hunk 11]
```diff
@@ -37,14 +37,27 @@ func getAllSnapshots(session *session.Session, region string, excludeAfter time.
...
+			snapshotIds = append(snapshotIds, snapshot.SnapshotId)
+		} else if !hasEBSSnapExcludeTag(snapshot) {
 			snapshotIds = append(snapshotIds, snapshot.SnapshotId)
 		}
 	}
 
 	return snapshotIds, nil
 }
 
+// hasEBSSnapExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
Comment (zackproser) [inline aws/snapshot.go:50]:
```suggestion
// hasEBSSnapExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

**Comment by zackproser** [comment]
This is looking very close @merkata, thanks for the great work! Left a couple NITs (one typo that was present in a couple of places, and some thoughts on the use of `aws.StringValue` and related methods for dereferencing pointers. Otherwise we should be good to get this merged shortly :+1:

**Comment by hongil0316** [comment]
Hey @merkata, we haven't heard back from you for quite some time. Do you plan to continue making progress on this PR? 
We've recently made some refactoring changes in the code and it may require a bit of conflict resolution.

**Comment by hongil0316** [comment]
Moved this change to this PR - https://github.com/gruntwork-io/cloud-nuke/pull/573.

**Pull request closed without merging**